### PR TITLE
(SERVER-763) Read client auth info from tk-authz when configured

### PR DESCRIPTION
**Pull request by camlow325**
This commit changes the Ruby request handler to read client
authentication info from trapperkeeper-authorization when tk-authz is
enabled via the `use-legacy-auth-conf: false` setting.  Support for the
pre-existing method of getting this info - via X-Client headers or an
SSL certificate, depending upon how the `master.allow-header-cert-info`
setting is configured - is preserved in the case that no information can
be derived from tk-authz.  tk-authz is presumed to provide client
authentication info via an `authorization` map within the Ring request
map.

### CODE REVIEW

**Comment by camlow325** [comment]
This won't pass Travis CI yet and isn't yet ready to merge because it depends upon a snapshot version of `trapperkeeper-authorization`.  In order for this PR to be ready to merge, We'll need to have merged https://github.com/puppetlabs/trapperkeeper-authorization/pull/17 and done a release of `trapperkeeper-authorization` that this PR can be bumped up to.  I've put this PR up in the meantime for early review.

Depending upon how long it takes for PR reviews to take place, I may add some additional docs-related commits with some updates to the language related to the `allow-header-cert-info` feature and `tk-auth` integration as well.  I'll add comments to this PR as I add any new docs commits.

**Comment by camlow325** [comment]
Added some updates to the `configuration.markdown` and `external_ssl_termination.markdown` docs in a separate commit - 5550e07.

**Comment by camlow325** [comment]
Rebased and updated to `trapperkeeper-authorization` 0.1.4 in d12449b.  Should be ready to review for possible merge now.

File: documentation/configuration.markdown [hunk 1]
```diff
@@ -150,67 +179,230 @@ profiler: {
...
+...
 ~~~
 
+If you want to add a rule but let the default rules from Puppet take
+precedence over your new rule, you should use a `sort-order` for the rule
+which is in the range from 601 to 998 (inclusive).
+
+Note that, for backward compatibility, the values of other configuration
+settings control the specific endpoints for which `trapperkeeper-authorization`
+is usedr:
```
Comment (nwolfe) [inline documentation/configuration.markdown]: s/usedr/used

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj [hunk 2]
```diff
@@ -113,13 +102,27 @@
...
+  [header-dn-name header-dn-val header-auth-name header-auth-val]
   (if (ssl-utils/valid-x500-name? header-dn-val)
-    {:client-cert-cn (ssl-utils/x500-name->CN header-dn-val)
-     :authenticated  (= "SUCCESS" header-auth-val)}
     (do
-      (if-not (nil? header-dn-val)
-        (log/errorf "The DN '%s' provided by the HTTP header '%s' is malformed."
+      (let [cn (ssl-utils/x500-name->CN header-dn-val)
+            authenticated (= "SUCCESS" header-auth-val)]
+        (log/debugf "CN '%s' provided by HTTP header '%s'"
```
Comment (dankreek) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: Logging++

File: documentation/configuration.markdown [hunk 3]
```diff
@@ -150,67 +179,230 @@ profiler: {
...
+
+* [`jruby-puppet.use-legacy-auth-conf`](#puppetserverconf) - Controls the
+ method to be used for authorizing access to the HTTP endpoints served by the
+ "master" service.  The applicable endpoints include those listed in the
+ [Puppet V3 HTTP API](https://docs.puppetlabs.com/puppet/4.2/reference/http_api/http_api_index.html#puppet-v3-http-api).
+
+ For a value of `true`, also the default value if not specified, authorization
+ will be done in core Ruby puppet-agent code via the legacy
+ [`auth.conf`](https://docs.puppetlabs.com/puppet/4.2/reference/config_file_auth.html)
+ file.  For a value of `false`, authorization will be done through via
```
Comment (nwolfe) [inline documentation/configuration.markdown]: s/through via/via

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj [hunk 4]
```diff
@@ -173,40 +176,68 @@
...
+                              header-dn-val
+                              header-auth-name
+                              header-auth-val)
+            (assoc :client-cert (header-cert header-cert-val)))
+        (do
+          (doseq [[header-name header-val]
+                  {header-dn-name header-dn-val
+                   header-auth-name header-auth-val
+                   header-client-cert-name header-cert-val}]
+            (if header-val
```
Comment (nwolfe) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: All the comprehension forms have special support for the common conditionals that follow, so you could collapse this `if` into the `doseq` like so:
```clojure
(doseq [[header-name header-val] {...}
        :when header-val]
...)
```

Might be worthwhile in this particular function with as many nesting levels as it has.

**Comment by nwolfe** [comment]
:+1: other than minor comments

**Comment by camlow325** [comment]
@nwolfe Addressed your PR feedback - typos and the doseq/when thing - in 9f79699.

File: src/clj/puppetlabs/services/request_handler/request_handler_core.clj [hunk 5]
```diff
@@ -173,40 +176,68 @@
...
+  Parameters:
+
+  * config - Map of configuration data
+  * request - Ring request containing client data"
+  [config request]
+  (if-let [authorization (:authorization request)]
+    {:client-cert    (ring-auth/authorized-certificate request)
+     :client-cert-cn (ring-auth/authorized-name request)
+     :authenticated  (true? (ring-auth/authorized-authentic? request))}
+    (let [headers (:headers request)
```
Comment (dankreek) [inline src/clj/puppetlabs/services/request_handler/request_handler_core.clj]: ring actually had a `get-header` function which allows you to grab headers by name directly from the request object, but this is totally fine with me as it is.

**Comment by dankreek** [comment]
:+1: Looks good, just had that one little "actually" comment that's not even remotely a blocker here

**Pull request merged**